### PR TITLE
Support zero-sized mappings

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
CTS

**Description**
Mapping an empty buffer should succeed still, and it should assume the mapped state.
That second part was missing.

**Testing**
Unlocks CTS

### CODE REVIEW

**Comment by kvark** [comment]
The workaround for the empty mapping is rather hacky. Open to better ideas :)

File: wgpu-core/src/device/life.rs [hunk 1]
```diff
@@ -674,28 +676,34 @@ impl<A: HalApi> LifetimeTracker<A> {
...
+                        Ok(ptr) => Some(ptr),
                         Err(e) => {
                             log::error!("Mapping failed {:?}", e);
-                            resource::BufferMapAsyncStatus::Error
+                            None
                         }
                     }
                 } else {
-                    resource::BufferMapAsyncStatus::Success
+                    ptr::NonNull::new(self.empty_mapping.as_mut_ptr())
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/life.rs:692]: Couldn't we just do `ptr::NonNull::new((&mut []).as_mut_ptr())`? We'd be recreating a zero sized slice on the other side, unless for some reason gecko needs to deref this pointer.

Comment (kvark) [reply wgpu-core/src/device/life.rs:692]: I don't think this is valid. A `*mut u8` isn't holding `[]` alive, so the `[]` will die instantly, and the pointer becomes invalid.

Comment (cwfitzgerald) [reply wgpu-core/src/device/life.rs:692]: That's the trick though, a zero size slice is pointing to nothing, which is why `&mut []` is of lifetime `'static`. The way that you can show this is that it's valid to create a zero-size slice to an empty vector which hasn't allocated on the heap at all.

Comment (kvark) [reply wgpu-core/src/device/life.rs:692]: that's tricky indeed :/

**Comment by kvark** [comment]
Weird, CTS fails on Windows but passes on macOS.

**Comment by cwfitzgerald** [comment]
Closing this as out of date.

**Pull request closed without merging**